File: src/main/scala/jetbrains/zeppelin/api/websocket/WebSocketAPI.scala
package jetbrains.zeppelin.api.websocket

import java.net.URI

import jetbrains.zeppelin.api
import jetbrains.zeppelin.api.ConnectionStatus.ConnectionStatus
import jetbrains.zeppelin.api._
import jetbrains.zeppelin.api.websocket.WebSocketApiProtocol._
import org.eclipse.jetty.websocket.api.Session
import org.eclipse.jetty.websocket.api.annotations._
import org.eclipse.jetty.websocket.client.{ClientUpgradeRequest, WebSocketClient}
import spray.json.{JsObject, _}

import scala.collection.mutable

//noinspection LoopVariableNotUpdated,RedundantBlock
@WebSocket(maxTextMessageSize = 64 * 1024)
class WebSocketAPI(uri: String) {
  private val client: WebSocketClient = new WebSocketClient()
  private val handlersMap: mutable.Map[String, MessageHandler] = mutable.Map()
  private val monitor = AnyRef
  var defaultHandler: MessageHandler = (_: ResponseMessage) => {
    println("Default Handler is called")
  }
  private var session: Session = _
  var status: api.ConnectionStatus.Value = ConnectionStatus.DISCONNECTED

  @OnWebSocketClose
  def onClose(statusCode: Int, reason: String): Unit = {
    status = ConnectionStatus.DISCONNECTED
  }

  @OnWebSocketError
  def onError(throwable: Throwable): Unit = {
    monitor.synchronized {
      status = ConnectionStatus.FAILED
      monitor.notifyAll()
    }
  }

  @OnWebSocketConnect
  def onConnect(session: Session) {
    this.session = session
    monitor.synchronized {
      status = ConnectionStatus.CONNECTED
      monitor.notifyAll()
    }
  }

  @OnWebSocketMessage
  def onMessage(msg: String) {
    val json = msg.parseJson.asJsObject
    val response = json.convertTo[ResponseMessage]
    val operationCode = response.op
    handlersMap.getOrElse(operationCode, defaultHandler).handle(response)
  }


  def connect(): ConnectionStatus = {
    val echoUri = new URI(uri)

    client.start()
    val request = new ClientUpgradeRequest()
    client.connect(this, echoUri, request)
    while (status == ConnectionStatus.DISCONNECTED) {
      waitMonitor()
    }
    status
  }

  private def waitMonitor() {
    monitor.synchronized {
      monitor.wait()
    }
  }

  def notifyMonitor(): Unit = {
    monitor.synchronized {
      monitor.notifyAll()
    }
  }

  def close() {
    client.stop()
  }


  def doRequestSync(requestMessage: RequestMessage, responseCode: String): JsObject = {
    if (status != ConnectionStatus.CONNECTED) throw SessionIsClosedException()


    var gotResponse = false
    var response: ResponseMessage = null

    registerHandler(responseCode, (result: ResponseMessage) => {
      monitor.synchronized {
        response = result
        gotResponse = true
        monitor.notifyAll()
      }
    })

    import WebSocketApiProtocol._
    val msg = requestMessage.toJson.toString()
    session.getRemote.sendString(msg)

    monitor.synchronized {
      while (!gotResponse) {
        monitor.wait()
      }
    }
    response.data
  }

  def doRequestAsync(requestMessage: RequestMessage, handlersMap: Map[String, MessageHandler]) {
    if (status != ConnectionStatus.CONNECTED) throw SessionIsClosedException()

    for (tuple <- handlersMap) registerHandler(tuple._1, tuple._2)

    import WebSocketApiProtocol._
    val msg = requestMessage.toJson.toString()
    session.getRemote.sendString(msg)
  }


  def registerHandler(op: String, handler: MessageHandler) {
    handlersMap += (op -> handler)
  }
}

File: zeppelin/implementation/src/org/intellij/plugin/zeppelin/api/remote/websocket/WebSocketAPI.kt
package org.intellij.plugin.zeppelin.api.remote.websocket

import com.intellij.openapi.diagnostic.Logger
import org.eclipse.jetty.websocket.api.Session
import org.eclipse.jetty.websocket.api.annotations.*
import org.eclipse.jetty.websocket.client.ClientUpgradeRequest
import org.eclipse.jetty.websocket.client.WebSocketClient
import org.intellij.plugin.zeppelin.models.ConnectionStatus
import org.intellij.plugin.zeppelin.models.SessionIsClosedException
import org.intellij.plugin.zeppelin.models.ZeppelinException
import org.intellij.plugin.zeppelin.utils.JsonParser
import java.net.URI

@Suppress("unused", "UNUSED_PARAMETER")
@WebSocket(maxTextMessageSize = 1024 * 1024)
class WebSocketAPI(address: String) {
    private val uri = "ws://$address/ws"
    private val logger: Logger = Logger.getInstance(WebSocketAPI::class.java)
    private val client: WebSocketClient = WebSocketClient()
    private val handlersMap: MutableMap<String, MessageHandler> = mutableMapOf()
    private val monitor = Object()
    private var defaultHandler: MessageHandler = object :
            MessageHandler {
        override fun handle(result: WsResponseMessage) {
            println("Default Handler is called. Operation: ${result.op}. Data: ${result.data}")
        }
    }
    var status = ConnectionStatus.DISCONNECTED
    private var session: Session? = null

    @OnWebSocketClose
    fun onClose(statusCode: Int, reason: String) {
        synchronized(monitor) {
            status = ConnectionStatus.DISCONNECTED
        }
    }

    @OnWebSocketError
    fun onError(throwable: Throwable) {
        synchronized(monitor) {
            status = ConnectionStatus.FAILED
            monitor.notifyAll()
        }
    }

    @OnWebSocketConnect
    fun onConnect(session: Session) {
        synchronized(monitor) {
            this.session = session
            status = ConnectionStatus.CONNECTED
            monitor.notifyAll()
        }
    }

    @OnWebSocketMessage
    fun onMessage(msg: String) {
        if (logger.isTraceEnabled) logger.trace("The message is handled $msg.")
        val response = JsonParser.fromStringObject(msg, WsResponseMessage::class.java)

        val operationCode = response.op
        handlersMap.getOrDefault(operationCode, defaultHandler).handle(response)
    }

    fun connect(): ConnectionStatus {
        val echoUri = URI(uri)
        client.start()
        val request = ClientUpgradeRequest()
        client.connect(this, echoUri, request)
        while (status == ConnectionStatus.DISCONNECTED) {
            waitMonitor()
        }
        return status
    }

    private fun waitMonitor() {
        synchronized(monitor) { monitor.wait() }
    }

    fun close() {
        client.stop()
    }

    fun doRequestSync(requestMessage: WsRequestMessage, responseCode: String): Any {
        if (status != ConnectionStatus.CONNECTED) throw SessionIsClosedException()
        var gotResponse = false
        var wsResponse: WsResponseMessage? = null

        val handleResult = object : MessageHandler {
            override fun handle(result: WsResponseMessage) {
                synchronized(monitor) {
                    wsResponse = result
                    gotResponse = true
                    monitor.notifyAll()
                }
            }
        }

        registerHandler(responseCode, handleResult)
        doRequestWithoutWaitingResult(requestMessage)

        synchronized(monitor) {
            while (!gotResponse && status == ConnectionStatus.CONNECTED) {
                monitor.wait()
            }
        }
        return wsResponse?.data ?: throw ZeppelinException()
    }

    fun doRequestWithoutWaitingResult(requestMessage: WsRequestMessage) {
        if (status != ConnectionStatus.CONNECTED) throw SessionIsClosedException()
        val msg: String = JsonParser.toJson(requestMessage)
        session?.remote?.sendString(msg)
    }

    fun registerHandler(op: String, handler: MessageHandler) {
        handlersMap.putAll(mapOf(op to handler))
    }

    fun doRequestAsync(requestMessage: WsRequestMessage) {
        doRequestWithoutWaitingResult(requestMessage)
    }
}